soc: riscv: telink_b91: add dfu support for Telink b91 platform

rikorsev opened the pull request:
This PR contains functionality required for correct work of Telink b91 platform with DFU subsystem and **mcuboot** bootloader.
It contains following changes:
- updates in default Kconfig and device tree for **tlsr9518adk80d** board
- fixes in UART driver to make it works properly after booting from **mcuboot** bootloader.
- add config for **smp_srv** sample.
- changes in linker script and Kconfig files for b91 soc.

Signed-off-by: Alexandr Kolosov <rikorsev@gmail.com>

lairdjm commented on samples/subsys/mgmt/mcumgr/smp_svr/boards/tlsr9518adk80d.overlay:
```diff
@@ -0,0 +1,27 @@
+ */
+
+/ {
+	chosen {
+		zephyr,uart-mcumgr = &uart0;
+		zephyr,shell-uart = &uart0;
+		zephyr,console = &uart1;
+	};
+};
+
+&uart0 {
+	status = "okay";
+	current-speed = <115200>;
+	pinctrl-0 = <&uart0_tx_pb2_default &uart0_rx_pb3_default>;
+	pinctrl-names = "default";
+};
```
Why are these being added to the smp_svr sample? The main UART should be part of the board dts not a specific application overlay

rikorsev replied:
Hi @lairdjm 

Thanks for remark. I've been removed uart0 from overlay but uart1 is still there, due to it is not part of dts for tlsr9518adk80d board. Shall I move uart1 to tlsr9518adk80d board as well?

Pretty the same approach as I've been used I saw here: https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/riscv/tlsr9518adk80d/tlsr9518adk80d.dts

Regards,
Alex

lairdjm replied:
It needs to be part of the board DTS yes, you cannot add an overlay for this board to every single sample application, that is untenable and a major maintenance burden

rikorsev replied:
@lairdjm 
Ok, got it. The .overlay currently removed. I guess dts stuff would be maintained by separate PR.

edersondisouza commented on drivers/flash/soc_flash_b91.c:
```diff
@@ -76,10 +85,41 @@ static int flash_b91_erase(const struct device *dev, off_t offset, size_t len)
+			page_nums -= FLASH_BLOCK_32K_PAGES;
+			offset += FLASH_BLOCK_32K_SIZE;
+
+		/* check if possible to erace single sector (4Kb) */
```
typo: "erace"

rikorsev replied:
Thanks! Fixed.

nvlsianpu reviewed (changes requested):
Flash driver extension is topic of other PR already https://github.com/zephyrproject-rtos/zephyr/pull/45940
Changes to different modules should go in dedicated commits. Tangled ones might go in one commit.

nvlsianpu commented on west.yml:
```diff
@@ -174,7 +174,8 @@ manifest:
         - crypto
     - name: mcuboot
-      revision: e86f575f68fdac2cab1898e0a893c8c6d8fd0fa1
+      url: https://github.com/mcu-tools/mcuboot
```
such reference might be used only for testing purpose. We are using zephyr-rtos/mcuboot fork here.
I will make its update right after the 3.1.0 release.

rikorsev replied:
@nvlsianpu 

Ok, thanks for info. Update to west.yml has been removed from PR.

rikorsev committed:
soc: riscv: telink_b91: add dfu related configurations for b91 platform

Modify Telink b91 config files and linker script with necessary config
settings required for DFU.

Signed-off-by: Alex Kolosov <rikorsev@gmail.com>

rikorsev committed:
boards: riscv: tlsr9518adk80d: add new flash partition to the board dts

- Remove flash_mspi node that contains old one flash partition
- Add new partition in flash node
- Add the partition into choosen
- Add led2, led3 into aliases

Signed-off-by: Alex Kolosov <rikorsev@gmail.com>

rikorsev committed:
drivers: serial: telink_b91: fix uart unknown state after boot from mcuboot

mcuboot may use a uart as logger. After application boot from mcuboot,
uart remains in some unknown state. It could leads to some unpredictable
uart behaviour. This commit fixes the issue by resetting of uart state
at the beginning of uart init.

Signed-off-by: Alex Kolosov <rikorsev@gmail.com>

rikorsev commented:
@nvlsianpu 

Thanks for review. I've been refactor it according to the remarks. Flash related stuff has been removed from the PR.
Rest code has been divided into smaller commits depends on touched module. 

Regards, 
Alex

fkokosinski commented on boards/riscv/tlsr9518adk80d/tlsr9518adk80d.dts:
```diff
@@ -86,6 +89,36 @@
+			reg = <0xf4000 0x0000b000>;
+		/* region <0xff000 0x1000> is reserved for
+		 * Telink B91 SDK's data (BLE MAC address etc.)
+		 */
```
Could you make this comment 80-cols wide?

rikorsev replied:
Done

fkokosinski reviewed (commented):
Minor nitpick, looks good otherwise.

lairdjm commented on samples/subsys/mgmt/mcumgr/smp_svr/boards/tlsr9518adk80d.conf:
```diff
@@ -0,0 +1 @@
+CONFIG_IMG_BLOCK_BUF_SIZE=256
```
I can't access the datasheet for this part, is there any particular reason that this needs to be set to 256 instead of the default 512?

rikorsev replied:
Hi @lairdjm 

Thanks for nice question. The answer is: b91 platform has 256 bytes flash page size.

Just for curiosity, I've check it once again and remove the config (which means that IMG_BLOCK_BUF_SIZE has default value 512). **mcumgr** image update operation hung immediately after start. The reason is that function **find_flash_total_size** (stream_flash.c) return false on the condition `if (ctx->buf_len > info->size)`. Also I see log message "Buffer size is bigger than page"

lairdjm replied:
@de-nordic Should this go in the boards folder in Kconfig.defconfig with an if around it or in the mcumgr sample? Seems like it would be applicable to many samples and end applications

rikorsev replied:
Hi @lairdjm, @de-nordic 

The config is moved into tlsr9518adk80d_defconfig with detailed comment. Please, take a look and approve the PR if everything is ok.

Regards,
Alex

rikorsev commented:
Hi @edersondisouza, @lairdjm, @nvlsianpu,

Please review the PR and approve it if there is no remarks from your side.

Regards,
Alex

lairdjm commented:
@de-nordic Can you check the comment in the sample application overlay?

rikorsev committed:
boards: riscv: tlsr9518adk80d: add IMG_BLOCK_BUF_SIZE to board defconfig

Add IMG_BLOCK_BUF_SZE config into tlsr9518adk80d_defconfig due to
limitations for buffer image writter. The buffer shall be less (multiple
of access alignment) or equal to flash page. tlsr9518adk80d boards use
external P25Q16 IC as flesh memory. Flash page size of the IC is 256
bytes. So that, it is maximum image writer buffer size for such kind of
boards.

Signed-off-by: Alexandr Kolosov <rikorsev@gmail.com>

edersondisouza reviewed (changes requested):
I guess that with this change you should simply remove the "[samples: smp_svr: add config for tlsr9518adk80d](https://github.com/zephyrproject-rtos/zephyr/pull/45973/commits/fa3202333f6a326423599c13a5845b60666d8acb)" patch. No need to add a file in one patch to just remove it in the next.
Please, make this `CONFIG_IMG_BLOCK_BUF_SIZE` a single patch on the board default config.

rikorsev commented:
Hi @edersondisouza 

Nice catch! You are completely right. Redundant patch has been removed.

Regards,
Alex

lairdjm commented:
@rikorsev this is no longer mcumgr related, please stop adding myself and de-nordic

rikorsev commented:
Hi @kgugala, @pgielda, @dcpleung, @kasunch, @mgielda, @tgorochowik, @nvlsianpu, fkokosinski

Please, review the PR and approve it if it is fine.

Regards,
Alex

rikorsev commented:
Hi @nvlsianpu,

Could you close the **change request**, as it already satisfied? It is blocking merge of the PR.

Regards,
Alex
